### PR TITLE
Better Defaults: Auto AD Selection for Newton Methods

### PR DESCRIPTION
**Pull request by avik-pal**
- [x] Cleanup the Poly Algorithms
- [x] Proper AD Selection

### CODE REVIEW

File: src/default.jl [hunk 1]
```diff
@@ -102,227 +102,141 @@ for more performance and then tries more robust techniques if the faster ones fa
...
-    if SciMLBase.successful_retcode(sol3)
-        return SciMLBase.build_solution(prob, alg, sol3.u, sol3.resid;
-                                        sol3.retcode, sol3.stats)
-    end
-
-    sol4 = SciMLBase.__solve(prob,  TrustRegion(;linsolve, precs, adkwargs...), args...; kwargs...)
-    if SciMLBase.successful_retcode(sol4)
-        return SciMLBase.build_solution(prob, alg, sol4.u, sol4.resid;
-                                        sol4.retcode, sol4.stats)
+# This version doesn't allocate all the caches!
```
Comment (ChrisRackauckas) [inline src/default.jl:141]: But it's dynamic dispatching everywhere.

File: src/default.jl [hunk 2]
```diff
@@ -102,227 +102,141 @@ for more performance and then tries more robust techniques if the faster ones fa
...
+        TrustRegion(; linsolve, precs, adkwargs...),
+        TrustRegion(; linsolve, precs,
+            radius_update_scheme = RadiusUpdateSchemes.Bastin, adkwargs...),
+    ]
+    filter!(!isnothing, algs)
+
+    sols = Vector{SciMLBase.NonlinearSolution}(undef, length(algs))
+
+    for (i, solver) in enumerate(algs)
+        sols[i] = SciMLBase.__solve(prob, solver, args...; kwargs...)
```
Comment (ChrisRackauckas) [inline src/default.jl:160]: This is dynamic

File: src/default.jl [hunk 3]
```diff
@@ -102,227 +102,141 @@ for more performance and then tries more robust techniques if the faster ones fa
...
+        if SciMLBase.successful_retcode(sols[i])
+            return SciMLBase.build_solution(prob, alg, sols[i].u, sols[i].resid;
+                sols[i].retcode, sols[i].stats, original = sols[i])
```
Comment (ChrisRackauckas) [inline src/default.jl:163]: Dynamic output

File: src/default.jl [hunk 4]
```diff
@@ -102,227 +102,141 @@ for more performance and then tries more robust techniques if the faster ones fa
...
+    1 ≤ current ≤ length(cache.caches) || error("Current choices shouldn't get here!")
+
+    current_cache = cache.caches[current]
+    while current ≤ length(cache.caches) #  && !all(terminated[current:end])
+        sol_tmp = solve!(current_cache)
+        SciMLBase.successful_retcode(sol_tmp) && break
+        current += 1
+        cache.current = current
+        current_cache = cache.caches[current]
     end
```
Comment (ChrisRackauckas) [inline src/default.jl:201]: No type stability

**Comment by ChrisRackauckas** [comment]
did you time it at all? It's completely dynamically dispatching and not type stable.

**Comment by avik-pal** [comment]
Yes not ready yet :sweat:

**Comment by avik-pal** [comment]
I will loop unroll that version to avoid adding sols for if we add anything to the cache later

**Comment by avik-pal** [comment]
Everything is type-stable to the extent that the output alg is a union type.

File: src/default.jl [hunk 5]
```diff
@@ -102,279 +104,176 @@ for more performance and then tries more robust techniques if the faster ones fa
...
-        end
-
-        
-
-        #cache.stats.nsteps += 1
+    1 ≤ current ≤ length(cache.caches) || error("Current choices shouldn't get here!")
+
+    current_cache = cache.caches[current]
+    while current ≤ length(cache.caches) #  && !all(terminated[current:end])
+        sol_tmp = solve!(current_cache)
```
Comment (ChrisRackauckas) [inline src/default.jl:233]: This will cause dynamic dispatches and allocations

**Comment by avik-pal** [comment]
@ChrisRackauckas this is ready to go in

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**